bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,324 @@
+                "select id as db_id, gloss_id, is_public, videofile from video_glossvideo",
+                f"{DATABASE_URL}",
+            ],
+            env=NEW_ENV,
+            shell=False,
+            check=True,
+            text=True,
+            stdout=f_obj,
+        )
+    with open(keys_file, "r") as f_obj:
+        nzsl_raw_keys_list = f_obj.readlines()
+    print(
+        f"{len(nzsl_raw_keys_list)} rows retrieved: {keys_file}",
+        file=sys.stderr,
+    )
+
+    # Separate out the NZSL db columns
+    # Write them to a dictionary, so we can do fast operations
+    for rawl in nzsl_raw_keys_list:
+        rawl = rawl.strip()
+        if not rawl:
+            continue
+        columns = rawl.split("|")
+        db_id = columns[0].strip()
+        gloss_id = columns[1].strip()
+        is_public = columns[2].strip().lower() == "t"
+        # 'videofile' data is also the key for S3
+        video_key = columns[3].strip()
+        # Each dictionary slot contains these values
+        this_nzsl_raw_keys_dict[video_key] = [db_id, gloss_id, is_public]
+
+    return this_nzsl_raw_keys_dict
+
+
+# Get the s3 keys present and absent from our NZSL keys
+def create_all_keys_dict(
+    this_s3_bucket_raw_keys_list, this_nzsl_raw_keys_dict, all_keys_file=ALL_KEYS_FILE
+):
+    print("Getting S3 keys present and absent from NZSL Signbank ...", file=sys.stderr)
+    nkeys_present = 0
+    nkeys_absent = 0
+    this_all_keys_dict = {}
+    for video_key in this_s3_bucket_raw_keys_list:
+        if video_key in this_nzsl_raw_keys_dict:
+            nkeys_present += 1
+            # Add 'Present' column to start
+            this_all_keys_dict[video_key] = [True] + this_nzsl_raw_keys_dict[video_key]
+        else:
+            nkeys_absent += 1
+            # Add 'Present' (absent) column to start
+            this_all_keys_dict[video_key] = [False, "", "", ""]
+    print(f"PRESENT: {nkeys_present} keys", file=sys.stderr)
+    print(f"ABSENT:  {nkeys_absent} keys", file=sys.stderr)
+
+    # Write all keys back to a file
+    with open(all_keys_file, "w") as f_obj:
+        for video_key, item_list in this_all_keys_dict.items():
+            outstr = (
+                f"{video_key}{CSV_DELIMITER}{CSV_DELIMITER.join(map(str, item_list))}\n"
+            )
+            f_obj.write(outstr)
+
+    return this_all_keys_dict
+
+
+# From the keys present in NZSL, get all their ACL information
+def output_csv(this_all_keys_dict):
+    print(f"Getting ACLs for keys from S3 ({AWS_S3_BUCKET}) ...", file=sys.stderr)
+
+    # CSV header
+    csv_header_list = [
+        "Video S3 Key",
+        "Postgres ID",
+        "Gloss ID",
+        "Signbank Public",
+        "Expected S3 Canned ACL",
+        "Actual S3 Canned ACL",
+    ]
+    print(CSV_DELIMITER.join(csv_header_list))
+
+    for video_key, [
+        is_present,
+        db_id,
+        gloss_id,
+        is_public,
+    ] in this_all_keys_dict.items():
+
+        if not is_present:
```

Comment:
issue: as previously discussed, we want to capture as much information as we can for every entry

for objects in the S3 bucket that don't have a corresponding entry in the database, we can still describe their "actual" ACL

also suggestion: I'd break this out into its own function - you effectively have two parts: 1. iterating over each item to create a csv, and 2. creating a csv row for an item. Those nicely form two functions:

```python
def build_csv_row(video_key, etc):
  is_public = False
  video_key = None

  if etc.is_present:
    video_key = # ...
    # ... etc ...

  canned_acl_actual = # ...
  # ... etc ...

  return [
    f"{video_key}",
    f"{db_id}",
    f"{gloss_id}",
    f"{is_public}",
    f"{canned_acl_expected}",
    f"{canned_acl_actual}",
  ]

def output_csv(this_all_keys_dict):
  for video_key, etc in this_all_keys_dict.items():
    print(CSV_DELIMITER.join(build_csv_row(video_key, etc)))
```

Reply:
Ok!

Reply:
The ACLS for absent members are now captured.
I'll address the output functionality next.

Reply:
Done.